### PR TITLE
ref: Discard unfinished spans when capturing transaction

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Removeing the unfinished child spans from a transaction when capturing the transaction.

## :bulb: Motivation and Context

Issue #1057 

## :green_heart: How did you test it?

Unit Test

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the CHANGELOG if needed
- [ ] I updated the docs if needed
- [ ] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## unreleased
+
+- ref: Discard unfinished spans when capturing transaction
```
Comment (brustolin) [inline CHANGELOG.md:5]:
```suggestion
- ref: Discard unfinished spans when capturing transaction #1066
```

File: Sources/Sentry/SentryTracer.m [hunk 2]
```diff
@@ -109,7 +109,12 @@ - (void)captureTransaction
 {
     NSArray *spans;
     @synchronized(_spans) {
-        spans = [_spans copy];
+        spans = [_spans
+            filteredArrayUsingPredicate:[NSPredicate
+                                            predicateWithBlock:^BOOL(id<SentrySpan> _Nullable span,
+                                                NSDictionary<NSString *, id> *_Nullable bindings) {
+                                                return span.isFinished;
+                                            }]];
```
Comment (philipphofmann) [inline Sources/Sentry/SentryTracer.m:117]: `h`: I think we have to do this in the client as someone could manually call `[hub captureEvent:transaction]`.

File: Sources/Sentry/SentryTracer.m [hunk 3]
```diff
@@ -109,7 +109,12 @@ - (void)captureTransaction
 {
     NSArray *spans;
     @synchronized(_spans) {
-        spans = [_spans copy];
+        spans = [_spans
+            filteredArrayUsingPredicate:[NSPredicate
```
Comment (philipphofmann) [inline Sources/Sentry/SentryTracer.m:113]: `l`: It would be nice to add a comment on why we do this.

File: Sources/Sentry/SentryTracer.m (hunk 2 above)
Comment (brustolin) [reply Sources/Sentry/SentryTracer.m:117]: No, the user doesn't have a reference to transaction. All it have is a span that cannot be used on captureEvent.

File: Sources/Sentry/SentryTracer.m (hunk 3 above)
Comment (brustolin) [reply Sources/Sentry/SentryTracer.m:114]: Sure, good idea.

File: Tests/SentryTests/SentrySpanTests.swift [hunk 4]
```diff
@@ -85,6 +87,20 @@ class SentrySpanTests: XCTestCase {
         XCTAssertEqual(serializedChild["parent_span_id"] as? String, span.context.spanId.sentrySpanIdString)
     }
     
+    func testFinishWithUnfinishedChild() {
```
Comment (bruno-garcia) [inline Tests/SentryTests/SentrySpanTests.swift:90]: L: What about describing the expected outcome of the test in its title?
```suggestion
    func testFinishWithUnfinishedSpanDropsSpan() {
```

**Review by bruno-garcia — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,7 +2,8 @@
 
 ## unreleased
 
-ref: Make calls to customSamplingContext nonnull #1061
+- ref: Discard unfinished spans when capturing transaction #1066
+- ref: Make calls to customSamplingContext nonnull #1061
```
Comment (bruno-garcia) [inline CHANGELOG.md:6]:
```suggestion
- ref: Discard unfinished spans when capturing transaction (#1066)
- ref: Make calls to customSamplingContext nonnull (#1061)
```

**Review by philipphofmann — APPROVED** [review]
LGTM

**Pull request merged**